Mention {mirai} in {future} docs (#111)
rstudio/promises: vignettes/promises_04_futures.Rmd

  %\VignetteEncoding{UTF-8}
---

```{css echo=FALSE}
.alert-success a, .alert-success a:visited {
  color: inherit;
  text-decoration: underline;
}
.alert code {
  color: inherit;
  background-color: inherit;
}
```

<div class="alert alert-success">

> While this article and others on this site focus on the `future` package, there's an up and coming package called [`mirai`](https://shikokuchuo.net/mirai/) that you may want to consider instead.

Reviewer: ```suggestion
While this article and others on this site focus on the `future` package, there's a much newer package called [`mirai`](https://shikokuchuo.net/mirai/) that you may want to consider instead.
```

Here are some factors to consider as you choose between the two.

> 1. The `future` package tries hard to automatically infer what variables and packages you need from the main R package, and makes those available to the child process. `mirai` doesn't try to do this for you; you need to pass in whatever data you need explicitly, and make `library()` calls explicitly inside of your inner code.

Reviewer: ```suggestion
1. The `future` package tries hard to automatically infer what variables and packages you need from the main R package, and makes those available to the child process. `mirai` doesn't try to do this for you; you need to pass in whatever data you need explicitly, and make package-namespaced calls explicitly inside of your inner code.
```

Reviewer: I know `library()` is probably clearer as a concept, but it's not good practice for users to be actually doing that inside a `mirai()` call!

> 2. `mirai` is very fast; it's much faster than `future` at starting up and has less per-task overhead.

Reviewer: ```suggestion
2. `mirai` is very fast; it's much faster than `future` at starting up and has less per-task overhead. `mirai` creates event-driven promises, whereas promises using `future` time-poll every 0.1 seconds. This makes `mirai` ideal where response times and latency are critical.
```

> 3. `future` is designed to be a general API for managing distributed programming across many types of computing backends. `mirai` is more narrowly scoped: it does support both local and distributed execution, but in the latter case, it only supports `mirai` workers.

Reviewer: ```suggestion
3. `future` is designed to be a general API supporting many types of distributed computing backends, and potentially offers more options. `mirai` on the other hand is its own system, whilst it does support both local and distributed execution.
```

> 4. `future` puts a big emphasis on complicated scenarios where futures launch other futures ("evaluation topologies"); mirai doesn't care about this.

Reviewer: ```suggestion
4. `mirai` is inherently queued, meaning it readily accepts more tasks than workers. This means you don’t need an equivalent of `future_promise()`. With `future` you need to manage cases where futures launch other futures ("evaluation topologies") upfront, whereas with `mirai` they will just work.
```

> </div>

Reviewer: ```suggestion
5. `mirai` supports task cancellation and the ability to interrupt ongoing tasks on the worker.
</div>
```
